src/stirling/source_connectors/socket_tracer/protocols/mqtt/parse.cc

```diff
@@ -724,6 +706,27 @@ ParseState ParseFrame(message_type_t type, std::string_view* buf, Message* resul
     return ParseState::kInvalid;
   }
 
+  // Updating the state for PUBLISH based on whether it is duplicate
+  if (control_packet_type == MqttControlPacketType::PUBLISH) {
+    if (result->dup) {
+      if (type == message_type_t::kRequest) {
+        state->send[std::tuple<uint32_t, uint32_t>(result->header_fields["packet_identifier"],
+                                                   result->header_fields["qos"])] += 1;
+      } else {
+        state->recv[std::tuple<uint32_t, uint32_t>(result->header_fields["packet_identifier"],
+                                                   result->header_fields["qos"])] += 1;
+      }
+    } else {
+      if (type == message_type_t::kRequest) {
+        state->send[std::tuple<uint32_t, uint32_t>(result->header_fields["packet_identifier"],
+                                                   result->header_fields["qos"])] = 0;
+      } else {
+        state->recv[std::tuple<uint32_t, uint32_t>(result->header_fields["packet_identifier"],
+                                                   result->header_fields["qos"])] = 0;
+      }
+    }
```

Comment:
Is it necessary to store a key in the state for `PUBLISH`'es that don't have any dups?

The `::operator[]` method will insert into the map if the key doesn't exist. So I believe the `+=` operator will work as expected without the non dup branch.